pkg/genericcontrolplane/aggregator.go

```diff
@@ -14,287 +14,184 @@ See the License for the specific language governing permissions and
-	{Group: "authorization.k8s.io", Version: "v1beta1"}:          {group: 17600, version: 9},
-	{Group: "certificates.k8s.io", Version: "v1"}:                {group: 17300, version: 15},
-	{Group: "certificates.k8s.io", Version: "v1beta1"}:           {group: 17300, version: 9},
-	{Group: "extensions", Version: "v1beta1"}:                    {group: 17150, version: 1}, // prioritize below networking.k8s.io, which contains the GA version of Ingress, the only resource remaining in extensions/v1beta1
-	{Group: "rbac.authorization.k8s.io", Version: "v1"}:          {group: 17000, version: 15},
-	{Group: "rbac.authorization.k8s.io", Version: "v1beta1"}:     {group: 17000, version: 12},
-	{Group: "rbac.authorization.k8s.io", Version: "v1alpha1"}:    {group: 17000, version: 9},
-	{Group: "apiextensions.k8s.io", Version: "v1"}:               {group: 16700, version: 15},
-	{Group: "coordination.k8s.io", Version: "v1"}:                {group: 16500, version: 15},
-	{Group: "coordination.k8s.io", Version: "v1beta1"}:           {group: 16500, version: 9},
-	{Group: "flowcontrol.apiserver.k8s.io", Version: "v1beta1"}:  {group: 16100, version: 12},
-	{Group: "flowcontrol.apiserver.k8s.io", Version: "v1alpha1"}: {group: 16100, version: 9},
-}
+	// Use withCluster here because each logical cluster can have a distinct set of APIs coming from its CRDs.
+	crdSpecs, _, _, err := downloader.Download(withCluster(s.CustomResourceDefinitions.GenericAPIServer.Handler.Director), "")
 
-func apiServicesToRegister(delegateAPIServer genericapiserver.DelegationTarget, registration autoregister.AutoAPIServiceRegistration, id string) []*v1.APIService {
-	apiServices := []*v1.APIService{}
+	mergedSpecs, err := builder.MergeSpecs(controlPlaneSpec, crdSpecs)
 
-	for _, curr := range delegateAPIServer.ListedPaths("") {
-		if curr == "/api/v1" {
-			apiService := makeAPIService(schema.GroupVersion{Group: "", Version: "v1"}, id)
-			registration.AddAPIServiceToSyncOnStart(apiService)
-			apiServices = append(apiServices, apiService)
-			continue
-		}
+	openAPIVersionedService, err := handler.NewOpenAPIService(mergedSpecs)
+	if err != nil {
+		klog.Errorf("Error while building OpenAPI schema: %v", err)
+	}
 
-		if !strings.HasPrefix(curr, "/apis/") {
-			continue
-		}
-		// this comes back in a list that looks like /apis/rbac.authorization.k8s.io/v1alpha1
-		tokens := strings.Split(curr, "/")
-		if len(tokens) != 4 {
-			continue
-		}
+	handler := &singlePathHandler{}
 
-		apiService := makeAPIService(schema.GroupVersion{Group: tokens[2], Version: tokens[3]}, id)
-		if apiService == nil {
-			continue
-		}
-		registration.AddAPIServiceToSyncOnStart(apiService)
-		apiServices = append(apiServices, apiService)
+	// In order to reuse the kube-openapi API as much as possible, we
+	// register the OpenAPI service in the singlePathHandler
+	err = openAPIVersionedService.RegisterOpenAPIVersionedService("/openapi/v2", handler)
```

Comment:
does this mean we merge openapi specs here in the hotpath?

Reply:
(Note, this isn't net net to our forks - I've just moved code around.)

Yes, merging is done on the fly.

Reply:
can you add a TODO for optimization. Am not sure we want that long-term. At least we might want some caching. Merging is heavy.